### PR TITLE
Added "empty" substitution command

### PR DESCRIPTION
**Pull request by javanegmond**
Implementation of my solution to #539.

Adds `_empty` and is really similar to `_arg`.

### CODE REVIEW

File: tools/roslaunch/src/roslaunch/substitution_args.py [hunk 1]
```diff
@@ -290,6 +315,7 @@ def resolve_args(arg_str, context=None, resolve_anon=True):
         'optenv': _optenv,
         'anon': _anon,
         'arg': _arg,
+		'empty': _empty,
```
Comment (gavanderhoorn) [inline tools/roslaunch/src/roslaunch/substitution_args.py:318]: Indentation.

**Comment by mikeferguson** [comment]
+1 to concept, looks like there is a fix needed to indentation though.

**Comment by dirk-thomas** [comment]
The `roslaunch` implementation with all the recursion and context switching is fragile at best. Other pull requests adding features like this (#296) have not been merged since it is difficult to judge if changes break the existing functionality.

Beside the actual code change it also requires significant documentation effort as well as adding unit tests to ensure that a new feature works correctly (and doesn't break existing functionality).

May be it would also help to "plan" what exact features are missing and should be added. `empty` seems to be a very specific test. Probably implementing a generic `compare` would be more flexible). May this could be discussed on the SIG roslaunch but it hasn't been active recently (http://wiki.ros.org/sig/roslaunch).

**Comment by piyushk** [comment]
@dirk-thomas @gavanderhoorn What's the correct way to proceed to get this feature (or equivalent) in. Here's what I gathered from Dirk's comment:
1. Implementing `compare` might be more powerful than `empty`. I agree with this.
2. The PR needs to also add unit tests for the new feature.
3. What's the best place to discuss this feature. The SIG looks like it's been dead for a while, and I'll assume that more recent contributors are not subscribed to it. I feel like this feature would be extremely useful.

**Comment by dirk-thomas** [comment]
There are two major concerns with this kind of PRs (there are multiple for roslaunch: e.g. #296, #382):
- Without having a clear "roadmap" what features are considered it is difficult to decide if they make sense the way they work, if they should be split into separate features or combined in a more generic feature and how they will interact with all the other functionality. This PR and the just closed #593 are a good example of that.
- Since the roslaunch code is rather fragile the chances for regressions are pretty big. And roslaunch is a core component for ROS so any regression would have a huge impact. On the other side there are not many tests (especially to ensure that more complex use cases are working). Therefore I hesitate to merge more complex changes (which sadly has the downside that interesting new features are not becoming available).

In order to move forward both concerns need to be worked on. But my major focus is currently on ROS 2 and therefore I can not spend any significant time on ROS 1 feature development. Therefore a group of people would need to actively discuss the "roadmap" aspect and determine which features should be integrated, which shouldn't and how they fit together. Additionally increasing the amount of tests which cover more use cases especially complex ones would help to ensure that changes don't break the existing behavior.

**Comment by piyushk** [comment]
@dirk-thomas Thanks for the detailed response!

I'll come back and take a look at some features I'm interested in when I have more time.

**Comment by dirk-thomas** [comment]
This feature should be possible once #707 gets merged. Therefore I will close this ticket.

**Pull request closed without merging**